File: py/todoosy/relative_date.py
"""
Todoosy Relative Date Parser

Mirror of `js/src/relative-date.ts`. Resolves keywords ("today", "tomorrow")
and relative offsets ("2 weeks", "in 3 days", "two months") to concrete ISO
dates. Resolution always happens at parse time; the persisted form on disk
is always a concrete date.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import date, datetime, timezone
from typing import Optional, Tuple

from dateutil.relativedelta import relativedelta

try:
    from zoneinfo import ZoneInfo  # Python 3.9+
except ImportError:  # pragma: no cover
    ZoneInfo = None  # type: ignore

from .types import Settings


@dataclass(frozen=True)
class ResolvedNow:
    """The user's "today" as an ISO YYYY-MM-DD string (in their TZ)."""
    today_iso: str


_NUMBER_WORDS = {
    'one': 1, 'two': 2, 'three': 3, 'four': 4, 'five': 5,
    'six': 6, 'seven': 7, 'eight': 8, 'nine': 9, 'ten': 10,
    'eleven': 11, 'twelve': 12, 'thirteen': 13, 'fourteen': 14, 'fifteen': 15,
    'sixteen': 16, 'seventeen': 17, 'eighteen': 18, 'nineteen': 19, 'twenty': 20,
    'twenty-one': 21, 'twenty-two': 22, 'twenty-three': 23, 'twenty-four': 24,
    'twenty-five': 25, 'twenty-six': 26, 'twenty-seven': 27, 'twenty-eight': 28,
    'twenty-nine': 29, 'thirty': 30, 'thirty-one': 31,
}

_TENS = {'twenty', 'thirty'}
_ONES_AFTER_TENS = {
    'one': 1, 'two': 2, 'three': 3, 'four': 4, 'five': 5,
    'six': 6, 'seven': 7, 'eight': 8, 'nine': 9,
}

_DAY_UNITS = {'d', 'day', 'days'}
_WEEK_UNITS = {'w', 'wk', 'wks', 'week', 'weeks'}
_MONTH_UNITS = {'month', 'months'}
_YEAR_UNITS = {'y', 'yr', 'yrs', 'year', 'years'}

_FUSED_OFFSET_RE = re.compile(r'^(\d+)([a-z]+)$', re.IGNORECASE)


def resolve_now(now: Optional[datetime] = None, settings: Optional[Settings] = None) -> ResolvedNow:
    instant = now if now is not None else datetime.now(tz=timezone.utc)
    if instant.tzinfo is None:
        # Treat naive as UTC for predictability.
        instant = instant.replace(tzinfo=timezone.utc)

    tz_name = settings.timezone if settings and settings.timezone else 'UTC'
    if ZoneInfo is None:
        local = instant.astimezone(timezone.utc).date()
    else:
        try:
            local = instant.astimezone(ZoneInfo(tz_name)).date()
        except Exception:
            local = instant.astimezone(timezone.utc).date()
    return ResolvedNow(today_iso=local.isoformat())


def _parse_iso(iso: str) -> date:
    y, m, d = (int(x) for x in iso.split('-'))
    return date(y, m, d)


def _add_days(iso: str, days: int) -> str:
    return (_parse_iso(iso) + relativedelta(days=days)).isoformat()


def _add_months(iso: str, months: int) -> str:
    # relativedelta is calendar-aware: Jan 31 + 1 month → Feb 28/29 (clipped).
    return (_parse_iso(iso) + relativedelta(months=months)).isoformat()


def _add_years(iso: str, years: int) -> str:
    return (_parse_iso(iso) + relativedelta(years=years)).isoformat()


def resolve_keyword(word: str, now: ResolvedNow) -> Optional[str]:
    w = word.lower()
    if w == 'today':
        return now.today_iso
    if w == 'tomorrow':
        return _add_days(now.today_iso, 1)
    return None


def _try_parse_number_word(parts: list[str], index: int) -> Optional[Tuple[int, int]]:
    """Returns (value, parts_consumed) or None."""
    if index >= len(parts):
        return None
    w0 = parts[index].lower()
    # Try compound form ("twenty one") before the simple lookup so it isn't
    # pre-empted by the value of "twenty".
    if w0 in _TENS and index + 1 < len(parts):
        w1 = parts[index + 1].lower()
        if w1 in _ONES_AFTER_TENS:
            total = _NUMBER_WORDS[w0] + _ONES_AFTER_TENS[w1]
            if total <= 31:
                return total, 2
    if w0 in _NUMBER_WORDS:
        return _NUMBER_WORDS[w0], 1
    return None


def _classify_unit(word: str) -> Optional[str]:
    w = word.lower()
    if w in _DAY_UNITS:
        return 'days'
    if w in _WEEK_UNITS:
        return 'weeks'
    if w in _MONTH_UNITS:
        return 'months'
    if w in _YEAR_UNITS:
        return 'years'
    return None


def _apply_offset(now: ResolvedNow, count: int, kind: str) -> str:
    if kind == 'days':
        return _add_days(now.today_iso, count)
    if kind == 'weeks':
        return _add_days(now.today_iso, count * 7)
    if kind == 'months':
        return _add_months(now.today_iso, count)
    if kind == 'years':
        return _add_years(now.today_iso, count)
    raise ValueError(f'unknown kind {kind!r}')


def _try_consume_from_suffix(parts: list[str], index: int) -> int:
    if index + 1 < len(parts):
        a = parts[index].lower()
        b = parts[index + 1].lower()
        if a == 'from' and b in ('now', 'today'):
            return 2
    return 0


def try_parse_relative(parts: list[str], index: int, now: ResolvedNow) -> Optional[Tuple[str, int]]:
    """Returns (iso_date, parts_consumed) or None."""
    if index >= len(parts):
        return None

    cursor = index
    consumed_in = 0
    if parts[cursor].lower() == 'in':
        consumed_in = 1
        cursor += 1
        if cursor >= len(parts):
            return None

    # Fused form: "2w", "2years", etc. Estimate units (m/h/d) handled elsewhere
    # — fused day form is intentionally rejected here so bare `2d` keeps
    # estimate semantics.
    fused = _FUSED_OFFSET_RE.match(parts[cursor])
    if fused:
        count = int(fused.group(1))
        kind = _classify_unit(fused.group(2))
        if kind is None or kind == 'days':
            return None
        after = cursor + 1
        suffix = _try_consume_from_suffix(parts, after)
        return _apply_offset(now, count, kind), consumed_in + 1 + suffix

    # Split form: number then unit
    digits = re.match(r'^(\d+)$', parts[cursor])
    if digits:
        count = int(digits.group(1))
        number_consumed = 1
    else:
        word = _try_parse_number_word(parts, cursor)
        if word is None:
            return None
        count, number_consumed = word

    unit_index = cursor + number_consumed
    if unit_index >= len(parts):
        return None
    kind = _classify_unit(parts[unit_index])
    if kind is None:
        return None
    if kind == 'days' and parts[unit_index].lower() == 'd':
        # Reject the bare `d` split form for clarity; require day/days.
        return None

    after = unit_index + 1
    suffix = _try_consume_from_suffix(parts, after)
    return _apply_offset(now, count, kind), consumed_in + number_consumed + 1 + suffix

File: js/src/relative-date.ts
/**
 * Todoosy Relative Date Parser
 *
 * Resolves keywords ("today", "tomorrow") and relative offsets ("2 weeks",
 * "in 3 days", "two months") to concrete ISO dates. The resolution always
 * happens at parse time; the persisted form on disk is always a concrete
 * date.
 */

import type { Settings } from './types.js';

export interface ResolvedNow {
  /** The user's "today" as an ISO YYYY-MM-DD string (in their TZ). */
  todayIso: string;
}

const NUMBER_WORDS: Record<string, number> = {
  one: 1, two: 2, three: 3, four: 4, five: 5,
  six: 6, seven: 7, eight: 8, nine: 9, ten: 10,
  eleven: 11, twelve: 12, thirteen: 13, fourteen: 14, fifteen: 15,
  sixteen: 16, seventeen: 17, eighteen: 18, nineteen: 19, twenty: 20,
  'twenty-one': 21, 'twenty-two': 22, 'twenty-three': 23, 'twenty-four': 24,
  'twenty-five': 25, 'twenty-six': 26, 'twenty-seven': 27, 'twenty-eight': 28,
  'twenty-nine': 29, thirty: 30, 'thirty-one': 31,
};

const TENS = new Set(['twenty', 'thirty']);
const ONES_AFTER_TENS: Record<string, number> = {
  one: 1, two: 2, three: 3, four: 4, five: 5,
  six: 6, seven: 7, eight: 8, nine: 9,
};

const DAY_UNITS = new Set(['d', 'day', 'days']);
const WEEK_UNITS = new Set(['w', 'wk', 'wks', 'week', 'weeks']);
const MONTH_UNITS = new Set(['month', 'months']);
const YEAR_UNITS = new Set(['y', 'yr', 'yrs', 'year', 'years']);

export function resolveNow(now: Date | undefined, settings: Settings | undefined): ResolvedNow {
  const instant = now ?? new Date();
  const tz = settings?.timezone ?? 'UTC';
  // Use Intl to get Y-M-D in the configured timezone. en-CA gives YYYY-MM-DD.
  let todayIso: string;
  try {
    todayIso = new Intl.DateTimeFormat('en-CA', {
      timeZone: tz,
      year: 'numeric', month: '2-digit', day: '2-digit',
    }).format(instant);
  } catch {
    // Invalid timezone → fall back to UTC
    todayIso = new Intl.DateTimeFormat('en-CA', {
      timeZone: 'UTC',
      year: 'numeric', month: '2-digit', day: '2-digit',
    }).format(instant);
  }
  return { todayIso };
}

function parseISOAsUTC(iso: string): Date {
  // iso is YYYY-MM-DD; build a UTC Date so arithmetic doesn't drift across DST.
  const [y, m, d] = iso.split('-').map(n => parseInt(n, 10));
  return new Date(Date.UTC(y, m - 1, d));
}

function dateToISO(d: Date): string {
  const y = d.getUTCFullYear();
  const m = String(d.getUTCMonth() + 1).padStart(2, '0');
  const day = String(d.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

function addDays(iso: string, days: number): string {
  const d = parseISOAsUTC(iso);
  d.setUTCDate(d.getUTCDate() + days);
  return dateToISO(d);
}

function addMonths(iso: string, months: number): string {
  // Calendar-aware: clip day if target month is shorter (Jan 31 + 1 month = Feb 28/29).
  const d = parseISOAsUTC(iso);
  const targetMonth = d.getUTCMonth() + months;
  const targetYear = d.getUTCFullYear() + Math.floor(targetMonth / 12);
  const normalizedMonth = ((targetMonth % 12) + 12) % 12;
  const desiredDay = d.getUTCDate();
  // Last day of target month: day 0 of next month.
  const lastDayOfTarget = new Date(Date.UTC(targetYear, normalizedMonth + 1, 0)).getUTCDate();
  const day = Math.min(desiredDay, lastDayOfTarget);
  return dateToISO(new Date(Date.UTC(targetYear, normalizedMonth, day)));
}

function addYears(iso: string, years: number): string {
  return addMonths(iso, years * 12);
}

/**
 * Resolve `today` or `tomorrow` (case-insensitive). Returns the ISO date or null.
 */
export function resolveKeyword(word: string, now: ResolvedNow): string | null {
  const w = word.toLowerCase();
  if (w === 'today') return now.todayIso;
  if (w === 'tomorrow') return addDays(now.todayIso, 1);
  return null;
}

/**
 * Try to parse a number-word starting at `index`. Returns count + parts consumed,
 * or null if no number-word found. Supports "twenty one" (two parts) and
 * "twenty-one" (single part with hyphen).
 */
function tryParseNumberWord(parts: string[], index: number): { value: number; partsConsumed: number } | null {
  if (index >= parts.length) return null;
  const w0 = parts[index].toLowerCase();
  // Try compound form first ("twenty one") so it doesn't get pre-empted by the
  // simple lookup of "twenty".
  if (TENS.has(w0) && index + 1 < parts.length) {
    const w1 = parts[index + 1].toLowerCase();
    if (w1 in ONES_AFTER_TENS) {
      const sum = NUMBER_WORDS[w0] + ONES_AFTER_TENS[w1];
      if (sum <= 31) return { value: sum, partsConsumed: 2 };
    }
  }
  if (w0 in NUMBER_WORDS) {
    return { value: NUMBER_WORDS[w0], partsConsumed: 1 };
  }
  return null;
}

/**
 * Identify a unit word and the relative-date "kind" it implies. Returns null
 * if the word is not a relative-date unit (estimate units `m`/`h`/`d` may
 * collide with `d`; the caller resolves that ordering — bare `2d` is parsed
 * as estimate before this is consulted).
 */
function classifyUnit(word: string): 'days' | 'weeks' | 'months' | 'years' | null {
  const w = word.toLowerCase();
  if (DAY_UNITS.has(w)) return 'days';
  if (WEEK_UNITS.has(w)) return 'weeks';
  if (MONTH_UNITS.has(w)) return 'months';
  if (YEAR_UNITS.has(w)) return 'years';
  return null;
}

function applyOffset(now: ResolvedNow, count: number, kind: 'days' | 'weeks' | 'months' | 'years'): string {
  switch (kind) {
    case 'days':   return addDays(now.todayIso, count);
    case 'weeks':  return addDays(now.todayIso, count * 7);
    case 'months': return addMonths(now.todayIso, count);
    case 'years':  return addYears(now.todayIso, count);
  }
}

const FUSED_OFFSET_RE = /^(\d+)([a-z]+)$/i;

/**
 * Try to consume a "from now" / "from today" suffix starting at `index`.
 * Returns the number of parts consumed (0, 2).
 */
function tryConsumeFromSuffix(parts: string[], index: number): number {
  if (index + 1 < parts.length) {
    const a = parts[index].toLowerCase();
    const b = parts[index + 1].toLowerCase();
    if (a === 'from' && (b === 'now' || b === 'today')) return 2;
  }
  return 0;
}

/**
 * Try to parse a relative-date expression starting at `index`. Returns the
 * resolved ISO date and parts consumed, or null if no relative date is
 * recognized at this position.
 *
 * Grammar (case-insensitive):
 *   ["in"] (digit | number-word) unit ["from now"|"from today"]
 *   ["in"] <fused>            where <fused> matches /^\d+(w|wk|wks|week|weeks|y|yr|yrs|year|years)$/
 *                             (note: bare 2d/2h/2m are estimates and parsed earlier)
 *
 * Returns null on no match. Does not throw. Bounded look-ahead.
 */
export function tryParseRelative(
  parts: string[],
  index: number,
  now: ResolvedNow,
): { date: string; partsConsumed: number } | null {
  if (index >= parts.length) return null;

  let cursor = index;
  let consumedIn = 0;
  if (parts[cursor].toLowerCase() === 'in') {
    consumedIn = 1;
    cursor++;
    if (cursor >= parts.length) return null;
  }

  // Try fused form first: "2w", "2years", etc. (single token, digits + alpha)
  const fused = parts[cursor].match(FUSED_OFFSET_RE);
  if (fused) {
    const count = parseInt(fused[1], 10);
    const kind = classifyUnit(fused[2]);
    if (kind === null) return null;
    // Day-fused (`2d`) is reserved for estimate; only weeks/months/years allowed fused.
    // (Estimate parser runs earlier and would have matched 2d/2h/2m already; if we
    // get here on `2d`, the estimate parser declined for some reason — still reject
    // to keep semantics one-way.)
    if (kind === 'days') return null;
    const after = cursor + 1;
    const suffixConsumed = tryConsumeFromSuffix(parts, after);
    return {
      date: applyOffset(now, count, kind),
      partsConsumed: consumedIn + 1 + suffixConsumed,
    };
  }

  // Split form: number then unit (each as separate part).
  // Number can be digits or a number-word (1-2 parts).
  let count: number;
  let numberPartsConsumed: number;
  const digitsMatch = parts[cursor].match(/^(\d+)$/);
  if (digitsMatch) {
    count = parseInt(digitsMatch[1], 10);
    numberPartsConsumed = 1;
  } else {
    const word = tryParseNumberWord(parts, cursor);
    if (!word) return null;
    count = word.value;
    numberPartsConsumed = word.partsConsumed;
  }

  const unitIndex = cursor + numberPartsConsumed;
  if (unitIndex >= parts.length) return null;
  const kind = classifyUnit(parts[unitIndex]);
  if (kind === null) return null;

  // Reject the day case in the *split* form when written as bare `d` ("2 d"):
  // that is a confusing form. Require "day"/"days" for clarity.
  if (kind === 'days' && parts[unitIndex].toLowerCase() === 'd') return null;

  const after = unitIndex + 1;
  const suffixConsumed = tryConsumeFromSuffix(parts, after);
  return {
    date: applyOffset(now, count, kind),
    partsConsumed: consumedIn + numberPartsConsumed + 1 + suffixConsumed,
  };
}
